File: AccessEye/AI/GemmaService.swift
//
//  GemmaService.swift
//  AccessEye
//
//  The REAL on-device describer: Gemma 3n (E2B) via Google's LiteRT-LM Swift API
//  with Metal GPU acceleration. Everything runs locally — the image never leaves
//  the device.
//
//  Why LiteRT-LM (not MediaPipe): the MediaPipe LLM Inference iOS API is
//  deprecated and never officially supported multimodal *image* input on iOS —
//  exactly the README §3 Risk #1. LiteRT-LM is Google's current engine, ships an
//  official Swift Package, supports image input, and runs on Metal. (Verified
//  against github.com/google-ai-edge/LiteRT-LM v0.13.0.)
//
//  This whole file is guarded by `#if canImport(LiteRTLM)` so the project still
//  builds if the package is ever removed; without it, calls report
//  `sdkNotInstalled` and the app falls back to `MockDescriber`.
//
//  To use it: add the model (.litertlm) and set AppConfig.useRealGemma = true.
//  See SETUP.md.
//

import UIKit

#if canImport(LiteRTLM)
import LiteRTLM

actor GemmaService: SceneDescriber {

    /// The engine is heavy to create, so we build it once and keep it.
    private var engine: Engine?

    func prepare() async throws {
        guard engine == nil else { return }

        guard let modelPath = ModelStore.usableModelPath else {
            throw DescriberError.modelFileMissing
        }

        do {
            // Metal GPU for both the language model and the vision encoder.
            // (This model's vision encoder *requires* the GPU backend; supplying
            // `visionBackend` is also what enables multimodal image input.)
            let config = try EngineConfig(
                modelPath: modelPath,
                backend: .gpu,
                visionBackend: .gpu,
                maxNumTokens: 1024,
                cacheDir: Self.cacheDirectory()
            )
            let engine = Engine(engineConfig: config)
            try await engine.initialize()
            self.engine = engine
        } catch {
            throw DescriberError.inferenceFailed(error.localizedDescription)
        }
    }

    func describe(image: UIImage, in language: Language) async throws -> String {
        try await prepare()
        guard let engine else { throw DescriberError.modelFileMissing }

        guard let jpeg = image.jpegData(compressionQuality: 0.85) else {
            throw DescriberError.inferenceFailed("Could not encode the captured image.")
        }

        do {
            let sampler = try SamplerConfig(topK: 40, topP: 0.95, temperature: 0.6)
            let conversationConfig = ConversationConfig(samplerConfig: sampler)

            // A fresh conversation per capture: each description is independent,
            // so earlier scenes don't bias the next one.
            let conversation = try await engine.createConversation(with: conversationConfig)

            // Image + the "describe for a blind user, in <language>" instruction.
            let message = Message(contents: [
                .imageData(jpeg),
                .text(Prompts.describeScene(in: language)),
            ])

            let response = try await conversation.sendMessage(message)
            let text = response.toString.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !text.isEmpty else {
                throw DescriberError.inferenceFailed("The model returned no text.")
            }
            return text
        } catch let error as DescriberError {
            throw error
        } catch {
            throw DescriberError.inferenceFailed(error.localizedDescription)
        }
    }

    /// A persistent directory for the engine's compiled GPU artifacts, so they
    /// survive across launches (faster, more reliable init than a temp dir).
    private static func cacheDirectory() -> String {
        let base = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
        let dir = base.appendingPathComponent("LiteRTLM", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.path
    }
}

#else

// Package not linked — provide a type with the same name so call sites compile.
actor GemmaService: SceneDescriber {
    func prepare() async throws {
        throw DescriberError.sdkNotInstalled
    }

    func describe(image: UIImage, in language: Language) async throws -> String {
        throw DescriberError.sdkNotInstalled
    }
}

#endif

File: android/app/src/main/java/gr/orestislef/accesseye/ai/GemmaService.kt
//
//  GemmaService.kt
//  AccessEye
//
//  The REAL on-device describer: Gemma 3n (E2B) via Google's LiteRT-LM Android
//  API. Everything runs locally — the image never leaves the device.
//
//  Why LiteRT-LM (not MediaPipe): the MediaPipe LLM Inference API is in
//  maintenance mode; LiteRT-LM is Google's current engine, consumes the same
//  `.litertlm` file as the iOS build, and supports multimodal image input.
//  Backends: CPU for text decode (GPU decode is slower and crash-prone on
//  Mali-class GPUs), GPU for the vision encoder (required for Gemma 3n).
//
//  The engine is heavy (~10s+ to initialize), so it is built exactly once,
//  guarded by a Mutex, always on a background dispatcher. Each describe() uses
//  a fresh Conversation so earlier scenes don't bias the next one.
//

package gr.orestislef.accesseye.ai

import android.content.Context
import android.graphics.Bitmap
import com.google.ai.edge.litertlm.Backend
import com.google.ai.edge.litertlm.Content
import com.google.ai.edge.litertlm.Contents
import com.google.ai.edge.litertlm.ConversationConfig
import com.google.ai.edge.litertlm.Engine
import com.google.ai.edge.litertlm.EngineConfig
import com.google.ai.edge.litertlm.SamplerConfig
import gr.orestislef.accesseye.model.Language
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream

class GemmaService(
    private val context: Context,
    private val store: ModelStore,
) : SceneDescriber {

    /** The engine is heavy to create, so we build it once and keep it. */
    @Volatile
    private var engine: Engine? = null

    /** Serializes initialization so concurrent prepare() calls build ONE engine. */
    private val initMutex = Mutex()

    override suspend fun prepare(): Unit = withContext(Dispatchers.IO) {
        initMutex.withLock {
            if (engine != null) return@withContext

            if (!store.isInstalled) {
                throw DescriberException.ModelFileMissing()
            }

            try {
                val engineConfig = EngineConfig(
                    modelPath = store.file.absolutePath,
                    // Text/decode on CPU: GPU decode is ~6x slower and unstable
                    // on Mali-G52-class GPUs. Vision MUST be GPU for Gemma 3n.
                    backend = Backend.CPU(),
                    visionBackend = Backend.GPU(),
                    maxNumTokens = 1024,
                    cacheDir = cacheDirectory(),
                )
                val newEngine = Engine(engineConfig)
                newEngine.initialize() // blocking, can take 10s+ — IO dispatcher
                engine = newEngine
            } catch (e: CancellationException) {
                throw e
            } catch (e: DescriberException) {
                throw e
            } catch (e: Exception) {
                // Retry (x3) lives in AppViewModel, like iOS.
                throw DescriberException.InferenceFailed(e.message ?: e.javaClass.simpleName)
            }
        }
    }

    override suspend fun describe(image: Bitmap, language: Language): String {
        prepare()
        val engine = engine ?: throw DescriberException.ModelFileMissing()

        return withContext(Dispatchers.Default) {
            // The vision encoder works at 768px — downscale first so we don't
            // waste time encoding pixels the model will never see.
            val png = toPngBytes(downscaled(image, MAX_IMAGE_SIDE))

            try {
                // A fresh conversation per capture: each description is
                // independent, so earlier scenes don't bias the next one.
                val conversation = engine.createConversation(
                    ConversationConfig(
                        samplerConfig = SamplerConfig(topK = 40, topP = 0.95, temperature = 0.6),
                    )
                )
                try {
                    // Image FIRST, text after (required for an accurate last token).
                    val contents = listOf(
                        Content.ImageBytes(png),
                        Content.Text(Prompts.describeScene(language)),
                    )
                    val reply = conversation.sendMessage(Contents.of(contents))
                    val text = reply.toString().trim()
                    if (text.isEmpty()) {
                        throw DescriberException.InferenceFailed("The model returned no text.")
                    }
                    text
                } finally {
                    conversation.close()
                }
            } catch (e: CancellationException) {
                throw e
            } catch (e: DescriberException) {
                throw e
            } catch (e: Exception) {
                throw DescriberException.InferenceFailed(e.message ?: e.javaClass.simpleName)
            }
        }
    }

    override fun close() {
        runCatching { engine?.close() }
        engine = null
    }

    /**
     * A persistent directory for the engine's compiled GPU artifacts, so they
     * survive across launches (faster, more reliable init than a temp dir).
     * Lives in noBackupFilesDir: the OS never purges it under storage pressure
     * (cacheDir can be wiped) and it stays out of device backups.
     */
    private fun cacheDirectory(): String =
        context.noBackupFilesDir.resolve("litertlm-cache").apply { mkdirs() }.absolutePath

    /** Scales [src] down so its longest side is at most [maxSide] (no-op if smaller). */
    private fun downscaled(src: Bitmap, maxSide: Int): Bitmap {
        val longest = maxOf(src.width, src.height)
        if (longest <= maxSide) return src
        val scale = maxSide.toFloat() / longest
        val w = (src.width * scale).toInt().coerceAtLeast(1)
        val h = (src.height * scale).toInt().coerceAtLeast(1)
        return Bitmap.createScaledBitmap(src, w, h, true)
    }

    private fun toPngBytes(bitmap: Bitmap): ByteArray =
        ByteArrayOutputStream().use { out ->
            if (!bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)) {
                throw DescriberException.InferenceFailed("Could not encode the captured image.")
            }
            out.toByteArray()
        }

    private companion object {
        const val MAX_IMAGE_SIDE = 768
    }
}
